doc/reference/trd-legal.md

```diff
@@ -0,0 +1,183 @@
+the rationale behind the license selection (dual-license, MIT or Apache2) and
+copyright policy (optional, and authors may retain copyright if desired). It
+further outlines how licensing and copyright shall be handled throughout
+projects under the Tock umbrella.
+
+Explicitly not discussed in this TRD are issues of trademark, the Tock brand,
+logos, or other non-code assets and artifacts.
+
+
+1. Introduction
+===============
+
+Tock’s goal is to provide a safe, secure, and efficient operating system for
+microcontroller-class devices. The Tock project further believes it is
+important to enable and support widespread adoption of safe, secure, and
+efficient software. Tock also seeks to be an open and inclusive project, and
+Tock welcomes contributions from any individuals or entities wishing to
+improve the safety, security, reliability, efficiency, or usability of Tock
+and the Tock ecosystem.
+
+The intent of these policies is to best satisfy the needs of all stakeholders
+in the Tock ecosystem.
+
+
+2. Licensing
+============
+
+All software artifacts under the umbrella of the Tock project are
+dual-licensed as Apache License, Version 2.0 (LICENSE-APACHE or
+http://www.apache.org/licenses/LICENSE-2.0) or MIT license (LICENSE-MIT or
+http://opensource.org/licenses/MIT).
+
+All contributions to Tock projects MUST be licensed under these terms.
+
+
+3. Copyright
+============
+
+Entities contributing resources to open-source projects often require
+attribution to recognize their efforts. Copyright notices are a common means to
+provide this. For downstream users, the license terms of Tock ensures
+unencumbered use.
+
+Copyrights in Tock projects are retained by their contributors. No
+copyright assignment is required to contribute to Tock projects.
+
+Artifacts in the Tock project MAY include explicit copyright notices.
+Substantial updates to an artifact MAY add additional copyright notices
+to an artifact. Copyright notices SHOULD NOT be removed by any party except
+the original copyright holder.
+
+For full authorship information, see the version control history.
+
+
+4. Implementation
+=================
+
+All code artifacts in Tock projects MUST include a license notice.
+
+Code artifacts MAY include copyright notice(s). Copyright notices SHOULD
+include a year. Newer copyright notices SHOULD be placed after existing
+copyright notices. If non-trivial updates are performed by an original
+copyright author, they MAY amend the year(s) indicated on their existing
+copyright statement or MAY add an additional copyright line, at their
+discretion.
+
+
+4.1 Format
+----------
+
+License and copyright information SHOULD appear near the top of a file.
+Generally, license and copyright information SHOULD be the first lines in a file
+unless necessary for technical (e.g. to accommodate the shebang in a shell
+script) or other reasons (e.g. importing a file with an existing copyright
+notice in a different location).
```

Comment:
We may want to use this "loophole" to avoid having the license specification be the first lines of a file which also includes a module-level documentation comment. In my (rather subjective) opinion, something like the following actually looks rather odd to people who expect module documentation to always be at the top of a file (which is motivated by Rust not allowing any actual code before such a parent-module doc comment). This would be a generous interpretation of _technical or other reasons_:

```
// Licensed under the Apache License, Version 2.0 or the MIT License.
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright Jane Doe <jdoe@example.org> 2019,2022

//! Provides userspace with access to a serial interface.
//!
//! Setup
//! -----
//!
//! You need a device that provides the `hil::uart::UART` trait.
```

I am aware this conflicts with the example provided below. Not trying to start a flamewar about code style here, just my 2ct.

Reply:
I think I interpret "necessary" different than you do. If we want the headers to go below module documentation, then we should state that explicitly.

Reply:
Oh, yeah, I agree with you. My comment is a convoluted way to ask the question "to me this looks weird, should we do this instead?". If we come to the conclusion that license specs should indeed follow doc-comments or even be part of them (at the end, which I've seen throughout the ecosystem), we might want to change the TRD.

Reply:
Conclusion: redundant since we're going to be more liberal with code imported from third-parties. Normal practice would be license header above module docs.